File: Java/src/main/java/org/chess/rendering/ConsoleRenderer.java
package org.chess.rendering;

import org.chess.domain.ChessPiece;
import org.chess.domain.GameState;
import org.chess.domain.PlayerColor;
import org.chess.domain.Position;

public class ConsoleRenderer {

    private static final String EMPTY_BOARD = """
                        
                 ╔═════╦═════╦═════╦═════╦═════╦═════╦═════╦═════╗
              8  ║     ║     ║     ║     ║     ║     ║     ║     ║
                 ╠═════╬═════╬═════╬═════╬═════╬═════╬═════╬═════╣
              7  ║     ║     ║     ║     ║     ║     ║     ║     ║
                 ╠═════╬═════╬═════╬═════╬═════╬═════╬═════╬═════╣
              6  ║     ║     ║     ║     ║     ║     ║     ║     ║
                 ╠═════╬═════╬═════╬═════╬═════╬═════╬═════╬═════╣
              5  ║     ║     ║     ║     ║     ║     ║     ║     ║
                 ╠═════╬═════╬═════╬═════╬═════╬═════╬═════╬═════╣
              4  ║     ║     ║     ║     ║     ║     ║     ║     ║
                 ╠═════╬═════╬═════╬═════╬═════╬═════╬═════╬═════╣
              3  ║     ║     ║     ║     ║     ║     ║     ║     ║
                 ╠═════╬═════╬═════╬═════╬═════╬═════╬═════╬═════╣
              2  ║     ║     ║     ║     ║     ║     ║     ║     ║
                 ╠═════╬═════╬═════╬═════╬═════╬═════╬═════╬═════╣
              1  ║     ║     ║     ║     ║     ║     ║     ║     ║
                 ╚═════╩═════╩═════╩═════╩═════╩═════╩═════╩═════╝
                    a     b     c     d     e     f     g     h
            """;

    private final char[] RENDER_BUFFER = new char[EMPTY_BOARD.length()];

    public String render(GameState gameState) {
        renderEmptyBoard();
        renderChessPieces(gameState);

        return new String(RENDER_BUFFER);
    }

    private void renderEmptyBoard() {
        System.arraycopy(EMPTY_BOARD.toCharArray(), 0, RENDER_BUFFER, 0, RENDER_BUFFER.length);
    }

    private void renderChessPieces(GameState gameState) {
        gameState.chessPiecePositions().forEach(this::renderChessPiece);
    }

    private void renderChessPiece(Position position, ChessPiece chessPiece) {
        final int horizontalOffsetForLabels = 5;
        final int squareWidth = 6;
        final int squareHeight = 2;
        final int horizontalOffsetWithinSquare = 4;

        // Rows on board are counted from bottom, but in the buffer from top.
        final int rowFromTop = 8 - position.row();

        final int initialEmptyLines = 1;
        final int line = initialEmptyLines + rowFromTop * squareHeight;
        final int offsetWithinLine = horizontalOffsetForLabels + (position.column() - 1) * squareWidth + horizontalOffsetWithinSquare;
        final int charsPerLine = 55;
        final int index = line * charsPerLine + offsetWithinLine;

        RENDER_BUFFER[index] = toChar(chessPiece);
    }

    private char toChar(ChessPiece chessPiece) {
        return switch (chessPiece.name()) {
            case PAWN -> chessPiece.color() == PlayerColor.WHITE ? '♙' : '♟';
            case ROOK -> chessPiece.color() == PlayerColor.WHITE ? '♖' : '♜';
            case KNIGHT -> chessPiece.color() == PlayerColor.WHITE ? '♘' : '♞';
            case BISHOP -> chessPiece.color() == PlayerColor.WHITE ? '♗' : '♝';
            case QUEEN -> chessPiece.color() == PlayerColor.WHITE ? '♕' : '♛';
            case KING -> chessPiece.color() == PlayerColor.WHITE ? '♔' : '♚';
        };
    }
}

File: Kotlin/src/main/kotlin/org/chess/rendering/ConsoleRenderer.kt
package org.chess.rendering

import org.chess.domain.*

class ConsoleRenderer {

    private val renderBuffer = CharArray(EMPTY_BOARD.length)

    fun render(gameState: GameState): String {
        renderEmptyBoard()
        renderChessPieces(gameState)
        return String(renderBuffer)
    }

    private fun renderEmptyBoard() {
        System.arraycopy(EMPTY_BOARD.toCharArray(), 0, renderBuffer, 0, renderBuffer.size)
    }

    private fun renderChessPieces(gameState: GameState) {
        gameState
            .chessPiecePositions
            .forEach(::renderChessPiece)
    }

    private fun renderChessPiece(position: Position, chessPiece: ChessPiece) {
        val horizontalOffsetForLabels = 5
        val squareWidth = 6
        val squareHeight = 2
        val horizontalOffsetWithinSquare = 4

        // Rows on board are counted from bottom, but in the buffer from top.
        val rowFromTop = 8 - position.row
        val initialEmptyLines = 1
        val line = initialEmptyLines + rowFromTop * squareHeight
        val offsetWithinLine =
            horizontalOffsetForLabels + (position.column - 1) * squareWidth + horizontalOffsetWithinSquare
        val charsPerLine = 55
        val index = line * charsPerLine + offsetWithinLine
        renderBuffer[index] = toChar(chessPiece)
    }

    private fun toChar(chessPiece: ChessPiece): Char {
        return when (chessPiece.name) {
            ChessPieceName.PAWN -> if (chessPiece.color === PlayerColor.WHITE) '♙' else '♟'
            ChessPieceName.ROOK -> if (chessPiece.color === PlayerColor.WHITE) '♖' else '♜'
            ChessPieceName.KNIGHT -> if (chessPiece.color === PlayerColor.WHITE) '♘' else '♞'
            ChessPieceName.BISHOP -> if (chessPiece.color === PlayerColor.WHITE) '♗' else '♝'
            ChessPieceName.QUEEN -> if (chessPiece.color === PlayerColor.WHITE) '♕' else '♛'
            ChessPieceName.KING -> if (chessPiece.color === PlayerColor.WHITE) '♔' else '♚'
        }
    }

    companion object {
        private const val EMPTY_BOARD = """
     ╔═════╦═════╦═════╦═════╦═════╦═════╦═════╦═════╗
  8  ║     ║     ║     ║     ║     ║     ║     ║     ║
     ╠═════╬═════╬═════╬═════╬═════╬═════╬═════╬═════╣
  7  ║     ║     ║     ║     ║     ║     ║     ║     ║
     ╠═════╬═════╬═════╬═════╬═════╬═════╬═════╬═════╣
  6  ║     ║     ║     ║     ║     ║     ║     ║     ║
     ╠═════╬═════╬═════╬═════╬═════╬═════╬═════╬═════╣
  5  ║     ║     ║     ║     ║     ║     ║     ║     ║
     ╠═════╬═════╬═════╬═════╬═════╬═════╬═════╬═════╣
  4  ║     ║     ║     ║     ║     ║     ║     ║     ║
     ╠═════╬═════╬═════╬═════╬═════╬═════╬═════╬═════╣
  3  ║     ║     ║     ║     ║     ║     ║     ║     ║
     ╠═════╬═════╬═════╬═════╬═════╬═════╬═════╬═════╣
  2  ║     ║     ║     ║     ║     ║     ║     ║     ║
     ╠═════╬═════╬═════╬═════╬═════╬═════╬═════╬═════╣
  1  ║     ║     ║     ║     ║     ║     ║     ║     ║
     ╚═════╩═════╩═════╩═════╩═════╩═════╩═════╩═════╝
        a     b     c     d     e     f     g     h
"""
    }
}
